### ISSUE TITLE
Cargo check fails to detect an error.

### ISSUE THREAD

**Issue by graylikeme** [rust-lang/rust#52484]
When running `cargo check` it fails to output an error that is detected with `cargo build` 
b_reg is u8 and running stable 1.27.1

```
error: attempt to shift right with overflow
  --> src\cpu\instructions.rs:39:36
   |
39 |             cpu.flags.half_carry = ((b_reg | 0x0F) - 1) >> 31 > 0;
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[deny(exceeding_bitshifts)] on by default
```

**Comment by graylikeme** [comment]
Opened in cargo project

**Comment by graylikeme** [comment]
This is the correct place after all

**Comment by estebank** [comment]
It seems like `cargo check` doesn't run the lints by default, which it arguably should.

**Comment by ehuss** [comment]
There are a few things that are not enabled for `--emit=rmeta`.  There is a general issue for this at #49292 that lists some of the things that are not checked.

**Comment by RalfJung** [comment]
@ehuss specifically for bitshift lints, there have been some fixes fairly recently (see https://github.com/rust-lang/rust/issues/70660). Could you try with the latest beta?

**Comment by RalfJung** [comment]
Sorry, pinged the wrong person -- @graylikeme could you test again with the latest beta?

Also it would help if you could distill this into a self-contained example so that we can reproduce the problem on the playground (https://play.rust-lang.org)

**Comment by ehuss** [comment]
Here is an example, it passes with `cargo check`, but fails `cargo build`, latest nightly `rustc 1.45.0-nightly (fa51f810e 2020-04-29)`.

```rust
fn main() {
    let a: u8 = 42;
    let b = ((a | 0x0F) - 1) >> 31 > 0;
    println!("{}", b);
}
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4998a76829fb44657cd8ed814fb2a191

**Comment by RalfJung** [comment]
Thanks! This looks like we don't run MIR optimizations with `cargo check`... https://github.com/rust-lang/rust/issues/70923 has some more concrete plans for this, so I am closing this here in favor of the other one even though that one is newer. I will post your example there.
